src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,1198 @@
+   }
+
+};
+
+void ActivateBuses(Steinberg::Vst::IComponent* component, const Steinberg::Vst::MediaType mediaType, const Steinberg::Vst::BusDirection direction, bool state)
+{
+   using namespace Steinberg;
+
+   Vst::BusInfo busInfo;
+   for(int i = 0, count = component->getBusCount(mediaType, direction); i < count; ++i)
+      if(component->getBusInfo(mediaType, direction, i, busInfo) == kResultOk &&
+         busInfo.flags & Vst::BusInfo::kDefaultActive)
+      {
+         component->activateBus(mediaType, direction, i, state ? 1 : 0);
+      }
+}
+
+wxString GetFactoryPresetsBasePath()
+{
+#ifdef __WXMSW__
+   WCHAR commonFolderPath[MAX_PATH];
+   if(SHGetFolderPathW(NULL, CSIDL_COMMON_APPDATA, NULL, 0, commonFolderPath) == S_OK)
+   {
+      return wxString(commonFolderPath) + "\\VST3 Presets\\";
+   }
+   return {};
+#elif __WXMAC__
+   return wxString("Library/Audio/Presets/");
+#elif __WXGTK__
+   return wxString("/usr/local/share/vst3/presets/");
+#endif
+}
+
+wxString GetPresetsPath(const wxString& basePath, const VST3::Hosting::ClassInfo& effectClassInfo)
+{
+   wxRegEx fixName(R"([\\*?/:<>|])");
+   wxString companyName = wxString (effectClassInfo.vendor()).Trim();
+   wxString pluginName = wxString (effectClassInfo.name()).Trim();
+
+   fixName.ReplaceAll( &companyName, { "_" });
+   fixName.ReplaceAll( &pluginName, { "_" });
+
+   wxFileName result;
+   result.SetPath(basePath);
+   result.AppendDir(companyName);
+   result.AppendDir(pluginName);
+   auto path = result.GetPath();
+
+   return path;
+}
+
+wxString GetFactoryPresetsPath(const VST3::Hosting::ClassInfo& effectClassInfo)
+{
+   return GetPresetsPath(
+      GetFactoryPresetsBasePath(),
+      effectClassInfo
+   );
+}
+
+}
+
+void VST3Effect::Initialize()
+{
+   using namespace Steinberg;
+
+   const auto& pluginFactory = mModule->getFactory();
+   if(auto effectComponent = pluginFactory.createInstance<Vst::IComponent>(mEffectClassInfo.ID()))
+   {
+      if(effectComponent->initialize(&LocalContext()) == kResultOk)
+      {
+         if(auto audioProcessor = FUnknownPtr<Vst::IAudioProcessor>(effectComponent))
+         {
+            mEffectComponent = effectComponent;
+            mAudioProcessor = audioProcessor;
+         }
+         else
+            //It's stated that "This interface must always be supported by audio processing plug-ins."
+            throw std::runtime_error("VST3 plugin does not provide audio processor interface");
+      }
+      else
+         throw std::runtime_error("Cannot initialize VST3 effect component");
+   }
+   else
+      throw std::runtime_error("Cannot create VST3 effect component");
+
+   //defaults
+   mSetup.processMode = Vst::kOffline;
+   mSetup.symbolicSampleSize = Vst::kSample32;
+   mSetup.maxSamplesPerBlock = mUserBlockSize;
+   mSetup.sampleRate = 44100.0;
+
+   ActivateBuses(mEffectComponent.get(), Vst::kAudio, Vst::kInput, true);
+   ActivateBuses(mEffectComponent.get(), Vst::kAudio, Vst::kOutput, true);
```

Comment:
I see that since the last version of this branch, you moved ActivateBuses out of ProcessInitialize into the constructor, and there is no longer a complementary de-activation.  That seems wrong, like a kind of leak.

Reply:
```
/** The bus should be activated by the host per default on instantiation (activateBus call is requested). By default a bus is inactive. */
kDefaultActive = 1 << 0,
```
Well, probably the funciton name is misleading here. Some plugins may have `kDefaultActive` flag set which means that we need to activate bus manually (this also means that some plugins do not requrire manual bus actiavtion at all), there is nothing said about bus deactivation. As I understand deactivation may be required when changes to bus arrangement are made (we don't support this feature yet, we always use default arrangement), so we don't need to make additional calls to `IAudioProcessor::activateBus`